rfcs/0007-api-errors.md

```diff
@@ -0,0 +1,81 @@
+documented errors to drive failure handling logic and we cannot change them without considering
+how it affects compatibility. It is important to have common agreement among contributors
+about the guidelines to expose new public errors.
+
+## Goals
+
+The goal of this RFC is to establish a set of principles for exposing public errors
+and a framework for maintaining them.
+
+## Principles
+
+- **Errors should not expose implementation.** The intent is to reserve 
+enough space for the implementation to change without affecting the public API.
+- **Errors should be prescriptive.** Ideally it should be clear to the user what they need to do
+to resolve an issue. For example, we should indicate whether a failure may be transient and
+an operation can be retried. It may not always be possible to have clear guidance for 
+unexpected errors, but ideally we can at least state whether the error is fatal.
+- **Prefer coarse error types.** A useful criteria to decide whether a new error type is needed 
+is whether the user would handle it differently than one of the existing types. 
+- **Use rich error messages.** Pack as much information as is needed to diagnose a problem
+into the error message, which will not be considered part of the public API.
+
+## Public API
+
+Publicly exposed errors will be maintained in `slatedb::Error`. Each new error type must be documented
+in this RFC. All public APIs will document the complete set of errors that may be returned. 
+
+Each exposed error type will expose a generic documentation string which contains a general description
+of the error and contains any prescriptive guidance (such as indicating whether a failed operation can
+be retried or is fatal).
+
+Each error  type will also expose a custom message field which is used to convey details about the specific
+instance of the error that was encountered. Unlike the error type, the message is not part of the 
+public API and can be changed without notice. The intent is to pack as much detail into the message
+for someone to understand the root cause, and if possible, what they should do to resolve it.
+
+```rust
+mod slatedb {
+    pub trait Error {
+        /// General description of the error and prescriptive actions
+        fn doc(&self) -> &str;
+        /// Custom message for specific instance details to identify the root cause 
+        fn message(&self) -> String;
+    }
+}
+```
+
+The complete set of public error types are listed below:
+
+- `Error::InvalidArgument`: One of the arguments passed by the user is invalid. This must be raised 
+before an operation is allowed to produce any side effects. The operation can be retried
+once the error has been addressed.
+- `Error::IoError`: There was an error accessing storage or other IO resources. The operation
+can be safely retried.
+- `Error::InternalError`: Reserved for unexpected cases such as invalid internal database states. 
```

Comment:
I feel like there are two cases here that we should distinguish:
- invalid in-memory state (e.g. a panic)
- invalid persisted state (e.g. db corruption)

Users will probably want to react to these differently. In the former case it might be reasonable to shutdown the db or process and restart it. In the latter case you may want to trigger an urgent alert or initiate some repair action (e.g. see https://github.com/slatedb/slatedb/pull/411).

Reply:
Given that #411 would introduce an OnCorruptionCallback, I wonder if we can use `InternalError` as the error type for all corruption errors, and instead add a separate parameter to the callback that defines the corruption type.

Reply:
The corruption call back currently looks like this. 
```rust
type OnCorruptionCallback = dyn Fn(CorruptionDetails) -> Result<(), CorruptionError>;
```
I could change it to return an `Error::InternalError` instead. 

> invalid in-memory state (e.g. a panic)
Can a panic be caught in rust? I'm assuming the source of the "invalid in-memory state" would be from assertions which would panic and not return an error. A panic feels like the correct thing to do if we can't trust memory. If so, there is no reason to distinguish between in-memory and persisted state corruption.